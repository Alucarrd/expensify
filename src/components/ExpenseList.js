import React from 'react';
import {connect} from 'react-redux';
import ExpenseListItem from './ExpenseListItem'
import selectExpenses from '../selectors/expenses';

export const ExpenseList = (props) => (
  <div>
    {
      props.expenses.length === 0 ? (
        <p>No Expenses</p>
      ) : (
        props.expenses.map((expense) => {
        return  <ExpenseListItem key={expense.id} {...expense} />;
        })
      )
    }


  </div>
);

//this map state to props will map the state you received from the connect (redux store)
//and map to props so it can be passed in to the component
const mapStateToProps = (state) => {
  return {
    expenses : selectExpenses(state.expenses, state.filters)
  };
}
//This will pass the state into the ExpenseList component
export default connect(mapStateToProps)(ExpenseList);
