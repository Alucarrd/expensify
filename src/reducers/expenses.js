const expenseReducerDefaultState = []

//const expenseReducer = (state = expenseReducerDefaultState, action) => {
export default (state = expenseReducerDefaultState, action) => {
  switch(action.type){
    case 'ADD_EXPENSE':
    //concat returns a new array instead of modifying current one
      return [...state, action.expense];

    case 'REMOVE_EXPENSE':
      //destructure id out
      return state.filter(({id}) => id !== action.id);
    case 'EDIT_EXPENSE':
      return state.map( (expense) => {
        if(expense.id === action.id){
          //this means, give me all the properties of the expense
          //and give me all the properties of the update object
          //last writes win
          return {
            ...expense,
            ...action.updates
          }
        }
        else{
          return expense;
        }

      })
    default:
      return state;
  }
}
//export default expenseReducer;
