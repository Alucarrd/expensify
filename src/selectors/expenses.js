
import moment from 'moment';

//get visible expense
//const getVisibleExpenses = (expenses, { text, sortBy, startDate, endDate }) => {
export default (expenses, { text, sortBy, startDate, endDate }) => {
  //destructure the filter
  return expenses.filter((expense) => {
    const createdAtMoment = moment(expense.createdAt);
    const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day'):true;
    const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day'):true;
    const textMatch = expense.description.toLowerCase().includes(text.toLowerCase()) ;

    //figure out if expense.description as the text variable string inside it
    //look into includes
    //convert both strings to lower case

    return startDateMatch && endDateMatch && textMatch
  }).sort( (a, b) => {
    if(sortBy === 'date'){
      return a.createdAt  < b.createdAt ? 1 : -1;
    }
    else if (sortBy === 'amount') {
      return a.amount < b.amount ? 1 : -1;

    }
  })

}
