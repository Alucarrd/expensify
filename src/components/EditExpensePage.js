import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { editExpense, removeExpense } from '../actions/expenses';

//Refactory to class based component so we can pull out inline function
//setup mapDispatchToProps, we need two props
//

//should render edit expense page
//should render edit expense -> spies
//should handle removeExpense
export class EditExpensePage extends React.Component{
  onSubmit = (expense) => {
    console.log("updated", expense)
    this.props.editExpense(this.props.expense.id, expense);
    this.props.history.push('/');
  }
  onRemove = () => {
    console.log("removing id:" + this.props.expense.id);
    this.props.removeExpense({id: this.props.expense.id});
    this.props.history.push('/');
  }
  render() {
    return (
      <div>
        <ExpenseForm
          expense={this.props.expense}
          onSubmit={this.onSubmit}
        />
        <button onClick={this.onRemove}>Remove</button>
      </div>
    )
  }
}

const mapDispatchToProps = (dispatch, props) => ({
  editExpense: (id, expense) => dispatch(editExpense(id, expense)),
  removeExpense: (data) => dispatch(removeExpense(data))
});
const mapStateToProps = (state, props) => {
  return {
    expense: state.expenses.find( (expense) => expense.id === props.match.params.id)
  }
}
export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
