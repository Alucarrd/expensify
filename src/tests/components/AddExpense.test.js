import React from 'react';
import { shallow } from 'enzyme';
//import the unconnected class
import { AddExpensePage } from '../../components/AddExpensePage';
import expenses from '../fixtures/expenses';


let AddExpense, history, wrapper;

beforeEach( () => {
  AddExpense=jest.fn();
  //our push will just be spying the push call for history
  history = { push: jest.fn() }
  wrapper = shallow(<AddExpensePage AddExpense={AddExpense} history={history} />);
});
test('Should render AddExpensePage correctly', () => {
  //setup spies

  expect(wrapper).toMatchSnapshot();

});


test('Should handle onSubmit for AddExpensePage', () => {
  //The beforeEach(callBack) will run before each test case runs
  // const onSubmit=jest.fn();
  // //our push will just be spying the push call for history
  // const history = { push: jest.fn() }
  // const wrapper = shallow(<AddExpensePage onSubmit={onSubmit} history={history} />);

  expect(wrapper).toMatchSnapshot();
  wrapper.find('ExpenseForm').prop("onSubmit")(expenses[1]);
  expect(history.push).toHaveBeenLastCalledWith('/');
  expect(AddExpense).toHaveBeenLastCalledWith(expenses[1]);
});
