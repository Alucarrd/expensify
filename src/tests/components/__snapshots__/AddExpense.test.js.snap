// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should handle onSubmit for AddExpensePage 1`] = `
<div>
  <h1>
    This is from my add expense
  </h1>
  <ExpenseForm
    onSubmit={[Function]}
  />
</div>
`;

exports[`Should render AddExpensePage correctly 1`] = `
<div>
  <h1>
    This is from my add expense
  </h1>
  <ExpenseForm
    onSubmit={[Function]}
  />
</div>
`;
